mitmproxy/proxy/mode_servers.py

```diff
@@ -333,6 +387,86 @@ async def handle_udp_connection(self, connection_id: tuple, handler: ProxyConnec
+
+    async def start(self) -> None:
+        assert self._server is None
+
+        wg_opts = dict()
+        host = self.mode.listen_host(ctx.options.listen_host)
+
+        if port := self.mode.listen_port(ctx.options.listen_port):
+            wg_opts["listen_port"] = port
+        if name := self.mode.wireguard_conf:
+            wg_opts["name"] = name
+        if peers := self.mode.wireguard_peers:
+            wg_opts["peers"] = peers
+
+        conf = mitmproxy_wireguard.WireGuardServerConf.new(**wg_opts)
+
+        self._server = await mitmproxy_wireguard.start_server(
+            host,
+            conf,
+            self.handle_tcp_connection,
+            self.handle_udp_datagram,
+        )
+        ctx.log.info(f"{self.log_desc} listening at {human.format_address(self.listen_addrs)}.")
+
+    async def stop(self) -> None:
+        assert self._server is not None
+
+        server = self._server
+        self._server = None
+
+        server.close()
+        await server.wait_closed()
+
+        ctx.log.info(f"Stopped {self.log_desc} at {human.format_address(self.listen_addrs)}.")
+
+    def make_top_layer(self, context: Context) -> Layer:
+        # is this correct?
+        return layers.modes.TransparentProxy(context)
+
+    async def handle_tcp_connection(self, stream: mitmproxy_wireguard.TcpStream) -> None:
+        connection_id = ("tcp", stream.get_extra_info("peername"), stream.get_extra_info("sockname"))
+        handler = WireGuardConnectionHandler(ctx.master, stream, ctx.options, self.mode)
+
+        handler.layer = self.make_top_layer(handler.layer.context)
+        with self.manager.register_connection(connection_id, handler):
+            await handler.handle_client()
+
+    def handle_udp_datagram(self, data: bytes, remote_addr: Address, local_addr: Address) -> None:
+        connection_id = ("udp", remote_addr, local_addr)
+
+        # TODO: implement UdpStream in mitmproxy_wireguard
+        if connection_id not in self.manager.connections:
+            # stream = mitmproxy_wireguard.UdpStream()
+            # handler = WireGuardConnectionHandler(ctx.master, stream, ctx.options, self.mode)
+            # self.manager.connections[connection_id] = handler
+            # asyncio.create_task(self.handle_udp_connection(connection_id, handler))
+            pass
+        else:
+            # handler = self.manager.connections[connection_id]
+            # stream = cast(mitmproxy_wireguard.UdpStream, handler.transports[handler.client].reader)
+            pass
+        # stream.feed_data(data, remote_addr)
+
+    async def handle_udp_connection(self, connection_id: tuple, handler: WireGuardConnectionHandler) -> None:
+        with self.manager.register_connection(connection_id, handler):
+            await handler.handle_client()
+
+    @property
+    def listen_addrs(self) -> tuple[Address, ...]:
+        # the WireGuard server listens on exactly one socket
+        return self._server.getsockname(),
```

Comment:
How is IPv4/IPv6 handled right now when you listen on `""`?

Reply:
The WireGuard server listens on on only one socket, so it's either IPv4 *or* IPv6, but never both.

I'm not sure if an empty string is even parsed into a `std::net::SocketAddr` successfully or not (but I'd guess not).

Reply:
`""` is our default listen_addr, so it would be nice if it works. 😅 
Anyhow - no need to do anything here for now, I was just wondering what happens.

Reply:
Right now the `listen_addr` is passed unmodified to `str::parse::<std::net::SocketAddr>()` here:
https://github.com/decathorpe/mitmproxy_wireguard/blob/0.1.0-alpha.3/src/lib.rs#L128-L131

I just checked, and an empty string is not an accepted host address for `SocketAddr`. What does the empty "" address actually do? If it just binds to `0.0.0.0` or `127.0.0.1`, I can use that as the default (or even make that configurable).

Reply:
`asyncio.create_server("", 8080, ...)` binds two sockets: One on `0.0.0.0`, and one on `::` (assuming both are available - otherwise it's just one). We should ultimately do the same for Rust so that things are consistent, but just defaulting to `0.0.0.0`  to get things running is fine with me